src/server.rs

```diff
@@ -119,51 +123,406 @@ impl Options {
+                    Ok(false)
+                } else {
+                    self.shutdown.set(Some(shutdown));
+                    Ok(true)
+                }
+            }
+            None => Ok(true),
+        }
+    }
+}
+
+struct Warn(&'static str);
+
+impl<T> FnOnce<T> for Warn {
+    type Output = ();
+
+    extern "rust-call" fn call_once(self, arg: T) -> Self::Output {
+        self.call(arg)
+    }
+}
+
+impl<T> FnMut<T> for Warn {
+    extern "rust-call" fn call_mut(&mut self, arg: T) -> Self::Output {
+        self.call(arg)
+    }
+}
+
+impl<T> Fn<T> for Warn {
+    extern "rust-call" fn call(&self, _: T) -> Self::Output {
+        warn!("{}", self.0)
+    }
+}
+
+struct AlwaysOk;
+
+impl<T> FnOnce<T> for AlwaysOk {
+    type Output = Result<(), ()>;
+
+    extern "rust-call" fn call_once(self, arg: T) -> Self::Output {
+        self.call(arg)
+    }
+}
+
+impl<T> FnMut<T> for AlwaysOk {
+    extern "rust-call" fn call_mut(&mut self, arg: T) -> Self::Output {
+        self.call(arg)
+    }
+}
+
+impl<T> Fn<T> for AlwaysOk {
+    extern "rust-call" fn call(&self, _: T) -> Self::Output {
+        Ok(())
+    }
+}
+
+type ShutdownStream = stream::Map<stream::Take<mpsc::UnboundedReceiver<oneshot::Sender<()>>>,
+                                  ShutdownSetter>;
+
+type ConnectionStream = stream::Map<unsync::mpsc::UnboundedReceiver<ConnectionAction>,
+                                    ConnectionWatcher>;
+
+struct ShutdownWatcher {
+    inner: stream::ForEach<stream::MapErr<stream::TakeWhile<stream::Merge<ShutdownStream,
+                                                                          ConnectionStream>,
+                                                            ShutdownPredicate,
+                                                            Result<bool, ()>>,
+                                          Warn>,
+                           AlwaysOk,
+                           Result<(), ()>>,
+}
+
+impl Future for ShutdownWatcher {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<(), ()> {
+        self.inner.poll()
+    }
+}
+
+/// Creates a future that completes when a shutdown is signaled and no connections are open.
+fn shutdown_watcher() -> (ConnectionTracker, Shutdown, ShutdownWatcher) {
+    let (shutdown_tx, shutdown_rx) = mpsc::unbounded::<oneshot::Sender<()>>();
+    let (connection_tx, connection_rx) = unsync::mpsc::unbounded();
+    let shutdown = Rc::new(Cell::new(None));
+    let connections = Rc::new(Cell::new(0));
+    let shutdown2 = shutdown.clone();
+    let connections2 = connections.clone();
+
+    let inner = shutdown_rx.take(1)
+        .map(ShutdownSetter { shutdown: shutdown })
+        .merge(connection_rx.map(ConnectionWatcher { connections: connections }))
+        .take_while(ShutdownPredicate {
+            shutdown: shutdown2,
+            connections: connections2,
+        })
+        .map_err(Warn("UnboundedReceiver resolved to an Err; can it do that?"))
```

Comment:
Is there any guidance from the docs about whether this can happen? If it's never supposed to resolve to an error, maybe this should be `unreachable!`?